### PR TITLE
[Bugfix] Fix panic when parsing zero-length or short byte slices

### LINKED ISSUE #7: Panic on pasing zero-length byte array to EncryptedMessage::from_bytes
When you pass a zero length (or <32 len) byte array to `EncryptedMessage::from_bytes`, it will panic

```
thread 'tokio-runtime-worker' panicked at /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/generic-ecies-0.1.0/src/lib.rs:545:63:
range end index 32 out of range for slice of length 0
```

https://github.com/LFDT-Lockness/generic-ecies/blob/c059ec86cac5b0b59be4fd2168a257d1598d2694/src/lib.rs#L545

### PR DESCRIPTION
**Pull request by PROWLERx15**
Fixes #7

### Changes

Added `WrongLen` variant to `DeserializeError` for indicating that the input buffer is too short to contain a valid message.

Replaced unchecked slice indexing (`&bytes[..compressed_len]`) with 
`bytes.get(..compressed_len).ok_or(DeserializeError::WrongLen)?`

Applied the same bounds-checked pattern to the remaining unchecked slice operations in `from_bytes` that would still panic on certain short inputs:
  - The uncompressed point fallback (`&bytes[..len]`)
  - The tag extraction (`&bytes[(l - tag_len)..]`), which could underflow when the buffer length is less than the tag length
  - The message slice (`&mut bytes[point_len..(l - tag_len)]`), same underflow issue
  
  
### Testing

Added two tests (`from_bytes_empty` and `from_bytes_too_short`) in the shared test macro so they run against both ciphersuites. These verify that short inputs return `DeserializeError::WrongLen` instead of panicking.

All existing tests pass with `cargo test --all-features`
New tests cover empty and too-short byte slices for both `curve25519xsalsa20hmac` and `curve25519aes128_cbchmac`

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -637,4 +649,7 @@ pub enum DeserializeError {
     /// Failed to read [`EncryptedMessage::ephemeral_key`]
     #[error("Ephemeral DH key is zero")]
     ZeroPoint(#[from] generic_ec::errors::ZeroPoint),
+    /// Input buffer is too short to contain a valid message
+    #[error("Input buffer is too short")]
+    WrongLen,
```
Comment (survived) [inline src/lib.rs]: Actually, `TooShort` would be a better name and would align well with the error message!

File: src/common.rs [hunk 2]
```diff
@@ -100,6 +100,26 @@ macro_rules! make_tests {
...
 
+            #[test]
+            fn from_bytes_empty() {
+                let mut bytes = vec![];
+                let result = super::EncryptedMessage::from_bytes(&mut bytes);
+                assert!(
+                    matches!(result, Err(crate::DeserializeError::WrongLen)),
+                    "expected WrongLen, got {result:?}",
+                );
+            }
```
Comment (survived) [inline src/common.rs]: could you update the test to test against every possible length of `bytes` that should result into the new error? to make sure that there's no edge-case that we could have missed.

**Review by survived — COMMENTED** [review]
Thanks for the contribution! I left some suggestions. 

Also, I think we can release the fix to crates-io. Can you do a major bump on the version & update the changelog please?

File: src/common.rs (hunk 2 above)
Comment (PROWLERx15) [reply src/common.rs]: Updated the test to iterate over every possible length, from `0 to min_valid_len - 1` (point + tag). 
For lengths shorter than `compressed_len` it asserts `TooShort` specifically, and for lengths in between it just asserts an error was returned since at that point the buffer is long enough to attempt point parsing but the data is invalid, so `InvalidPoint` is the correct error there.

File: src/lib.rs (hunk 1 above)
Comment (PROWLERx15) [reply src/lib.rs]: renamed `WrongLen` to `TooShort` everywhere

**Comment by PROWLERx15** [comment]
@survived when you say major bump, do you mean `0.1.0 -> 1.0.0` or `0.1.0 -> 0.2.0` ?

**Comment by survived** [comment]
It's `0.1.0 -> 0.2.0`

File: src/common.rs [hunk 3]
```diff
@@ -100,24 +100,31 @@ macro_rules! make_tests {
...
+                    );
+                    // Inputs shorter than the compressed point length must return TooShort.
+                    // Longer inputs may fail with InvalidPoint instead, which is also correct.
+                    if len < compressed_len {
+                        assert!(
+                            matches!(result, Err(crate::DeserializeError::TooShort)),
+                            "expected TooShort for input length {len}, got {result:?}",
+                        );
+                    }
+                }
```
Comment (survived) [inline src/common.rs]: I think a better structure for this test would be:

1. `bytes = valid_point.to_bytes(true) + empty_tag` (by `+` I mean concatenation here), then test that `EncryptedMessage::from_bytes(&bytes[0..i])` fails with `TooShort` error for each `i in 0..(compressed_len + tag_len)`
2. `bytes = valid_point.to_bytes(false) + empty_tag`, then test that `EncryptedMessage::from_bytes(&bytes[0..i])` fails with `TooShort` error for each `i in uncompressed_len..(uncompressed_len + tag_len)`

In this way we test every possible scenario in which we return this error.

File: src/common.rs [hunk 4]
```diff
@@ -100,6 +100,33 @@ macro_rules! make_tests {
...
```
Comment (maurges) [inline src/common.rs]: 1. This allocates a new vector on every iteration. You could instead make an array of max len once, and subslice it in the cycle
2. You should put a valid point as the first bytes in this array so that invalid point errors never happen

File: src/common.rs (hunk 3 above)
Comment (PROWLERx15) [reply src/common.rs]: i have restructured the test into two parts as you suggested. compressed case iterates `0..(compressed_len + tag_len)` and uncompressed iterates `uncompressed_len..(uncompressed_len + tag_len)`, both using a valid point so we actually hit the `TooShort` path every time

File: src/common.rs (hunk 4 above)
Comment (PROWLERx15) [reply src/common.rs]: fixed both, single buffer allocated upfront and subsliced in the loop, and using a valid random point as the prefix so `InvalidPoint` never triggers

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,9 @@
+## v0.2.0
+
+- Fixed panic in `EncryptedMessage::from_bytes` when given zero-length or too-short input (#7)
+- Added `DeserializeError::TooShort` variant for graceful error handling on short inputs
+- Replaced unchecked slice indexing with bounds-checked access in `from_bytes`
```
Comment (maurges) [inline CHANGELOG.md]: No need to be so verbose. You should leave only the first line of those three

**Comment by maurges** [comment]
Looks good otherwise

File: CHANGELOG.md (hunk 5 above)
Comment (PROWLERx15) [reply CHANGELOG.md]: removed lines 2 & 3

File: src/common.rs [hunk 6]
```diff
@@ -100,6 +100,47 @@ macro_rules! make_tests {
...
+                        "expected TooShort for compressed input length {i}, got {result:?}",
+                    );
+                }
+
+                // Test with uncompressed point: valid_point ++ zero_tag
+                let uncompressed_bytes = valid_point.to_bytes(false);
+                let mut buf = Vec::with_capacity(uncompressed_len + tag_len);
+                buf.extend_from_slice(&uncompressed_bytes);
+                buf.extend_from_slice(&vec![0u8; tag_len]);
+                for i in uncompressed_len..(uncompressed_len + tag_len) {
```
Comment (maurges) [inline src/common.rs]: Why does this not start from zero? Or even from `compressed_len`?

File: src/common.rs [hunk 7]
```diff
@@ -100,6 +100,47 @@ macro_rules! make_tests {
...
+                let uncompressed_len = generic_ec::Point::<E>::serialized_len(false);
+                let tag_len =
+                    cipher::generic_array::GenericArray::<u8, crate::MacSize<super::S>>::default()
+                        .len();
+
+                // Test with compressed point: valid_point ++ zero_tag
+                let compressed_bytes = valid_point.to_bytes(true);
+                let mut buf = Vec::with_capacity(compressed_len + tag_len);
+                buf.extend_from_slice(&compressed_bytes);
+                buf.extend_from_slice(&vec![0u8; tag_len]);
```
Comment (maurges) [inline src/common.rs]: Here and below, you don't need to create an intermediate vector to extend from. You could either `resize(new_computed_len_with_a_better_variable_name, 0)`, or copy from the tag itself, which could be generated with `GenericArray::default`. I leave it to you to choose which fits better here

File: src/common.rs [hunk 8]
```diff
@@ -100,6 +100,47 @@ macro_rules! make_tests {
                 assert_eq!(pubkey, pubkey_);
             }
 
+            #[test]
+            fn from_bytes_too_short() {
+                let mut rng = rand_dev::DevRng::new();
+                let valid_point = generic_ec::Point::generator()
+                    * generic_ec::NonZero::<generic_ec::Scalar<E>>::random(&mut rng);
```
Comment (maurges) [inline src/common.rs]: I don't see a need to generate a random point. A zero or a generator would do as fine

File: src/common.rs [hunk 9]
```diff
@@ -100,6 +100,47 @@ macro_rules! make_tests {
...
+
+                let compressed_len = generic_ec::Point::<E>::serialized_len(true);
+                let uncompressed_len = generic_ec::Point::<E>::serialized_len(false);
+                let tag_len =
+                    cipher::generic_array::GenericArray::<u8, crate::MacSize<super::S>>::default()
+                        .len();
```
Comment (maurges) [inline src/common.rs]: All of those len machinations may not be necessary at all: if you have a point, you can `p.to_bytes(true/false).to_vec()`, and then append a tag to the resulting vector

File: src/lib.rs [hunk 10]
```diff
@@ -540,26 +540,38 @@ impl<'m, S: Suite> EncryptedMessage<'m, S> {
...
-                Err(e1) => {
-                    let len = generic_ec::Point::<S::E>::serialized_len(false);
-                    match generic_ec::Point::<S::E>::from_bytes(&bytes[..len]) {
-                        Ok(point) => (len, point),
-                        Err(e2) => return Err(DeserializeError::InvalidPoint(e1, e2)),
-                    }
+        let (point_len, ephemeral_key) = match generic_ec::Point::<S::E>::from_bytes(
+            bytes
+                .get(..compressed_len)
+                .ok_or(DeserializeError::TooShort)?,
```
Comment (maurges) [inline src/lib.rs]: Nitpick: it's easy to lose the track when a match scrutinee is 4 lines long. I think it would be better to separate `bytes.get...` into a variable. Same idea below

**Review by maurges — CHANGES_REQUESTED** [review]
Please avoid force-pushing unless absolutely necessary. They make it impossible to review changes incrementally

File: src/common.rs (hunk 8 above)
Comment (PROWLERx15) [reply src/common.rs]: switched to `Point::generator()`

File: src/common.rs (hunk 7 above)
Comment (PROWLERx15) [reply src/common.rs]: made the change. i'm now using `GenericArray::default()` and extending the buf directly

File: src/common.rs (hunk 9 above)
Comment (PROWLERx15) [reply src/common.rs]: yeah you're right, removed all the explicit length variables. just doing `point.to_bytes(true).to_vec()` and extending with the tag now, lengths fall out of `buf.len()`

File: src/lib.rs (hunk 10 above)
Comment (PROWLERx15) [reply src/lib.rs]: agreed, extracted both `bytes.get(...)` calls into `compressed_slice` and `uncompressed_slice variables`. much easier to read the match arms now

File: src/common.rs (hunk 6 above)
Comment (PROWLERx15) [reply src/common.rs]: good catch, started it from `compressed_len` now

**Comment by PROWLERx15** [comment]
> Please avoid force-pushing unless absolutely necessary. They make it impossible to review changes incrementally


@maurges apologies. I had missed signing a commit message, which caused the CI to fail. I had to force-push to correct that mistake. I’ll make sure to avoid force-pushing going forward unless it’s necessary.

**Comment by maurges** [comment]
@survived why did we do a major bump here? Because of the new error variant? Should we add a `#[non_exhaustive]` to avoid this being breaking in the future?

**Comment by survived** [comment]
Yes @maurges, because PR changes `enum` that is exposed in public API. I usually go with `pub struct Error(Reason); enum Reason { ... }` approach so the error type can evolve freely. `#[non_exhaustive]` also works (as long as we don't change existing variants of enum), but I guess if we change this error type, we need to change all error types.

**Comment by maurges** [comment]
Created #12 to remember this discussion

**Review by maurges — APPROVED** [review]
